### ISSUE TITLE
Resource extras can not be updated

### ISSUE THREAD

**Issue by wardi** [ckan/ckan#2158]
```
ckanapi action package_create name=testres
ckanapi action resource_create package_id=testres url=example newfield=first
```
^ this creates a new package with one resource including one extra "newfield"

In the database tables resource and resource_revision both contain:
```
   url   |        extras         
---------+-----------------------
 example | {"newfield": "first"}
```

```
ckanapi action resource_update id=<resource id> url=changed-url newfield=changed-newfield
```
^ this successfully updates the url, but the newfield extra remains unchanged.

In the database we see in resource (newfield was not updated):
```
     url     |        extras         
-------------+-----------------------
 changed-url | {"newfield": "first"}
```

But in resource_revision the correct data is stored (new revision on top is correct):
```
     url     |              extras              
-------------+----------------------------------
 changed-url | {"newfield": "changed-newfield"}
 example     | {"newfield": "first"}
```

Maybe vdm-related? There is also a code in package_update that avoids creating revisions that could be the cause.

**Comment by benjaminlaot** [comment]
Hi @wardi ! 
Can you give me a lead, so I can work on the problem on my side?

**Comment by clementmouchet** [comment]
Hi @wardi, @rossjones & and @davidread could you give us a lead to this fix?

**Comment by clementmouchet** [comment]
Sorry to chase you again on that @wardi, I could spend some time trying to fix this but so far I've not even been able to put my finger on the actual cause of the problem! Could you give us a lead?

**Comment by wardi** [comment]
@clementmouchet sorry, I don't understand. It looks like this problem or one like it has existed for a very long time so bisecting isn't helping. @davidread has created a test that reproduces the problem so the next thing I would do is step in to the failing test with a debugger.

**Comment by amercader** [comment]
Trying to wrap my head around this, one thing that seems suspicious is that we are using the [JsonDictType](https://github.com/ckan/ckan/blob/master/ckan/model/types.py#L67) for the `extras` field, which seems to be a variation of [this example](http://docs.sqlalchemy.org/en/rel_0_9/core/custom_types.html#marshal-json-strings) on the SQLAlchemy docs. In there it says:

> Note that the ORM by default will not detect “mutability” on such a type - meaning, in-place changes to values will not be detected and will not be flushed. Without further steps, you instead would need to replace the existing value with a new one on each parent object to detect changes. Note that there’s nothing wrong with this, as many applications may not require that the values are ever mutated once created. For those which do have this requirement, support for mutability is best applied using the sqlalchemy.ext.mutable extension - see the example in Mutation Tracking.

**Comment by wardi** [comment]
@amercader ok, try `res.extras = res.extras` just before saving

**Comment by amercader** [comment]
@wardi you mean here?

```diff
diff --git a/ckan/lib/dictization/model_save.py b/ckan/lib/dictization/model_save.py
index 4ba4788..878389e 100644
--- a/ckan/lib/dictization/model_save.py
+++ b/ckan/lib/dictization/model_save.py
@@ -50,7 +50,7 @@ def resource_dict_save(res_dict, context):
         del obj.extras[delete_me]
 
     obj.state = u'active'
-
+    obj.extras = obj.extras
     session.add(obj)
     return obj
```

I doesn't seem to make much difference but I see where you are coming from

**Comment by wardi** [comment]
Yeah, that what I was thinking -- maybe there's some sneaky code that checking if the thing you're assigning has changed. Could always go deeper: Assign some other object first, and/or pass to dict() before assigning.

Or better yet let's just not treat .extras as mutable in the code that uses it, because that doesn't work.

**Comment by amercader** [comment]
Replacing the extras with a brand new dict did the trick and solved the two issues. Fix is on @davidread's branch: #2159

**Comment by benjaminlaot** [comment]
Nice job !!

**Comment by wardi** [comment]
fixed in #2159
